ext/liquid_c/block.c

```diff
@@ -290,32 +318,87 @@ static VALUE internal_block_body_parse(block_body_t *body, parse_context_t *pars
+typedef struct block_body_yield_tag_args {
+    block_body_t *body;
+    serialize_parse_context_t *serialize_context;
+    tag_markup_header_t *current_tag;
+} block_body_yield_tag_args_t;
+
+static VALUE block_body_try_yield_tag(VALUE uncast_args)
 {
-    if (body->compiled) {
-        rb_raise(rb_eRuntimeError, "Liquid::C::BlockBody is already compiled");
-    }
+    block_body_yield_tag_args_t *args = (block_body_yield_tag_args_t *)uncast_args;
+    tag_markup_header_t *current_tag = args->current_tag;
+
+    serialize_parse_context_enter_tag(args->serialize_context, current_tag);
+    VALUE tag_name = rb_utf8_str_new(tag_markup_header_name(current_tag), current_tag->tag_name_len);
+    VALUE markup = rb_utf8_str_new(tag_markup_header_markup(current_tag), current_tag->markup_len);
+    return rb_yield_values(2, tag_name, markup);
 }
 
-static void ensure_intermediate_not_parsing(block_body_t *body)
+static VALUE block_body_rescue_yield_tag(VALUE uncast_args, VALUE exception)
 {
+    block_body_yield_tag_args_t *args = (block_body_yield_tag_args_t *)uncast_args;
+
+    serialize_parse_context_exit_tag(args->serialize_context, &args->body->as.serialize.document_body_entry,
+                                     args->current_tag);
+    rb_exc_raise(exception);
+}
+
+static VALUE block_body_parse_from_serialize(block_body_t *body, VALUE tokenizer_obj, VALUE parse_context_obj)
+{
+    assert(body->from_serialize);
+    assert(is_serialize_parse_context_p(parse_context_obj));
+
     ensure_intermediate(body);
+    if (body->as.serialize.parse_context != parse_context_obj) {
+        rb_raise(rb_eArgError, "Liquid::C::BlockBody#parse called with different parse context");
+    }
 
-    if (body->as.intermediate.code->parsing) {
-        rb_raise(rb_eRuntimeError, "Liquid::C::BlockBody is in a incompletely parsed state");
+    serialize_parse_context_t *serialize_context;
+    SerializeParseContext_Get_Struct(parse_context_obj, serialize_context);
+
+    body->as.serialize.document_body_entry = serialize_context->current_entry;
+
+    tag_markup_header_t *current_tag = serialize_context->current_tag;
+    while (current_tag) {
+        bool tag_unknown = TAG_UNKNOWN_P(current_tag);
+
+        if (tag_unknown) {
+            block_body_yield_tag_args_t yield_args = {
+                .body = body,
+                .serialize_context = serialize_context,
+                .current_tag = current_tag
+            };
+            return rb_rescue(block_body_try_yield_tag, (VALUE)&yield_args, block_body_rescue_yield_tag, (VALUE)&yield_args);
+        } else {
+            VALUE tag_name = rb_utf8_str_new(tag_markup_header_name(current_tag), current_tag->tag_name_len);
+            VALUE markup = rb_utf8_str_new(tag_markup_header_markup(current_tag), current_tag->markup_len);
+
+            VALUE tag_class = rb_funcall(tag_registry, intern_square_brackets, 1, tag_name);
+            assert(RTEST(tag_class));
```

Comment:
I think this should raise? Seems technically possible that tag could get removed between the time the template is parsed, cached (for a long time) & deserialized.

Maybe using `Block.unknown_tag`?

Reply:
:+1: Good point. On a unrelated note, should `block.c` be renamed to `block_body.c` since it actually contains the code for `Liquid::C::BlockBody`?

Reply:
yes, block_body.c would be a more appropriate name for this file.